src/libraries/System.Text.RegularExpressions/src/dotnet-regular-expressions.md

```diff
@@ -0,0 +1,237 @@
+
+Performance is important and we welcome optimizations so long as they preserve the public contract.
+
+## Extensibility
+
+Key types have significant protected surface area. This is probably not intended as an general extensibility point, but rather as a detail of implementing saving a compiled regex to disk. Saving to disk is implemented by saving an assembly containing three types, one that derives from each of `Regex`, `RegexRunnerFactory`, and `RegexRunner`. This mechanism accounts for all the protected methods (and even protected fields) on these classes. If we were designing them today, we would likely more carefully limit their public surface, and possibly not rely on derived types.
+
+Protected members are part of the public API which cannot be broken, so they may potentially make some future optimizations more difficult. In particular, we must keep them stable in order to remain compatible with regexes saved by .NET Framework.
+
+`RegexCompiler` is abstract for a different reason: to share implementation between `RegexLWCGCompiler` and `RegexAssemblyCompiler`: it based around a field of type `System.Reflection.Emit.ILGenerator` and has protected utility methods and fields to work with it. External extension by derivation of `RegexCompiler` would likely be clumsy as it contains knowledge of `RegexLWCGCompiler` and `RegexAssemblyCompiler`.
+
+## Key types - General
+
+### Regex (public)
+
+* Represents an executable regular expression with some utility static methods
+* Several protected fields and methods but no derived classes exist in this implementation (see [Extensibility](#Extensibility) section above).
+* Constructor sets `RegexCode` using `RegexParser` and `RegexWriter`; then, if `RegexOptions.Compiled`, compiles and holds a `RegexRunnerFactory` and clears `RegexCode`; these steps only need to be done once for this `Regex` object
+* Various public entry points converge on `Run()` which uses the held `RegexRunner` if any; if none or in use, creates another with the held `RegexRunnerFactory` if any; if none, interprets with held `RegexCode`
+* All static methods (such as `Regex.Match`) attempt to find a pre-existing `Regex` object for the requested pattern and options in the `RegexCache`. This is legitimate, since after construction, `Regex` options are thread-safe. If there is a cache hit, execution can begin immediately.
+
+### RegexOptions (public)
+
+* `RightToLeft` is supported throughout, but as the less common case it is less optimized.
+* `ExplicitCapture` is off by default: this is relevant to performance, as often patterns contain parentheses as a useful grouping mechanism, for example `(something){1,3}` is easier to type than the non capturing form `(?:something){1,3}`. Because explicit capture is off by default, the engine in this case will capture `something` even if it was not needed.
```

Comment:
Why did you call out only these two options?

Reply:
They particularly relate to optimization - the others are just details. Although, there are optimizations related to ignore case..

Reply:
Ok, then it'd be good to say that, i.e. state why you chose to include just these two when talking about RegexOptions.  There are other options relevant to performance as well, though.